internal/pkg/util/fs/files/passwd.go

```diff
@@ -65,6 +65,7 @@ func Passwd(path string, home string, uid int, customLookup UserGroupLookup) (co
 
 	sylog.Verbosef("Creating template passwd file and injecting user data: %s", path)
 	userExists := false
+	uid = int(pwInfo.UID)
```

Comment:
I'm a little confused here...

What's the circumstance in which `pwInfo.UID` now does not equal the `uid` we were passed in?

Reply:
I found exactly one such circumstance (so far): it arose during the ContainerAsFakeroot e2e test (in the ACTIONS group).

Reply:
Hmmm. It's odd that it would only come out in a FUSE related test... and a single one at that. It feels like something is wrong with the cherry-picked change, or alternatively that how it's working in Apptainer isn't really meshing with changes we have.

Looking through I can't see anything obvious. I'm going to need to spend a while on this to understand exactly what's going on.... hopefully tomorrow 🤞

Reply:
Possibly worth noting is that the test in question makes use of `e2e/testdata/sshfs.def`, and we've seen before (in the very PR having to do with Passwd(),namely #1531 ) an edge-case that was surfaced by precisely this definition file. The edge-case in question was, at the time, in the situation where the container lacked `/bin/bash`, but the run environment did have `/bin/bash` installed, the shell field of the root user in the container passwd file would get overwritten with `/bin/bash` creating a bricked container. This is obviously not _that_ problem, but it might be worthwhile to think what else might be special about this container...

Reply:
Having looked at it properly now, I think it's better to address this further up in the code. The root cause is that we are re-writing the `root` line for custom `--home` after your revamp of the passwd handling. Therefore it's the root passwd info we actually need, so we should pull that out by target[UG]ID = 0 if `--fakeroot` mode is active.

Apptainer doesn't re-write the `root` line, so it doesn't need to do this.

This should work...

```diff
diff --git a/internal/pkg/runtime/engine/singularity/prepare_linux.go b/internal/pkg/runtime/engine/singularity/prepare_linux.go
index 47437dae2..e4e92e2b5 100644
--- a/internal/pkg/runtime/engine/singularity/prepare_linux.go
+++ b/internal/pkg/runtime/engine/singularity/prepare_linux.go
@@ -130,7 +130,7 @@ func (e *EngineOperations) PrepareConfig(starterConfig *starter.Config) error {
 
        uid := e.EngineConfig.GetTargetUID()
        gids := e.EngineConfig.GetTargetGID()
-       useTargetIDs := false
+       useTargetIDs := e.EngineConfig.GetFakeroot()
 
        if os.Getuid() == 0 && (uid != 0 || len(gids) > 0) {
                starterConfig.SetTargetUID(uid)
diff --git a/internal/pkg/util/fs/files/passwd.go b/internal/pkg/util/fs/files/passwd.go
index 00050d065..0a23c2b97 100644
--- a/internal/pkg/util/fs/files/passwd.go
+++ b/internal/pkg/util/fs/files/passwd.go
@@ -65,7 +65,6 @@ func Passwd(path string, home string, uid int, customLookup UserGroupLookup) (co
 
        sylog.Verbosef("Creating template passwd file and injecting user data: %s", path)
        userExists := false
-       uid = int(pwInfo.UID)
        for i, line := range lines {
                if line == "" {
                        continue
```

Minimally verified against the `fuse_mount` tests only...

```
--- PASS: TestE2E/PAR/ACTIONS/fuse_mount (16.19s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostDaemonAsRoot (0.17s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostAsRoot (0.16s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerDaemonAsRoot (0.15s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerAsRoot (0.16s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostDaemonAsUser (0.18s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostAsUser (0.15s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerDaemonAsUser (0.15s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerAsUser (0.15s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostDaemonAsUserNamespace (0.22s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostAsUserNamespace (0.16s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerDaemonAsUserNamespace (0.16s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerAsUserNamespace (0.15s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostDaemonAsFakeroot (0.16s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/HostAsFakeroot (0.15s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerDaemonAsFakeroot (0.17s)
                --- PASS: TestE2E/PAR/ACTIONS/fuse_mount/ContainerAsFakeroot (0.15s)
```